File: Benchmarking/src/gen_8_30.scala
import java.lang.System.nanoTime
import scala.collection.immutable.Vector
object Gen extends App {
  case class s0 (
    p0: Vector[Vector[Double]],
    p1: Vector[Vector[Double]],
    p2: Vector[Vector[Double]]
  )
  case class s1 (
    p0: s0,
    p1: s0
  )
  @noinline
  def f0(v0: Vector[Vector[s1]], v1: Double): Double = {
    val v7: Vector[s1] = v0(0)
    val v3: s1 = v7(0)
    val v2: s0 = v3.p1
    val v17: Vector[Vector[Double]] = v2.p2
    val v27: Vector[Double] = v17(0)
    val v43: Double = v27(0)
    var v47: Double = v43
    val v91: Double = v47 / v47
    val v64: Double = v27(0)
    val v90: Double = v64 / v91
    v47 = v43
    v90
  }
  def benchmark(): Unit = {
    val v0: Vector[Vector[s1]] = Vector(Vector(s1(s0(Vector(Vector(0.0), Vector(1.0), Vector(2.0)), Vector(Vector(3.0)), Vector(Vector(4.0))), s0(Vector(Vector(5.0), Vector(6.0), Vector(7.0)), Vector(Vector(8.0)), Vector(Vector(9.0))))), Vector(s1(s0(Vector(Vector(10.0), Vector(11.0), Vector(12.0)), Vector(Vector(13.0)), Vector(Vector(14.0))), s0(Vector(Vector(15.0), Vector(16.0), Vector(17.0)), Vector(Vector(18.0)), Vector(Vector(19.0))))), Vector(s1(s0(Vector(Vector(20.0), Vector(21.0), Vector(22.0)), Vector(Vector(23.0)), Vector(Vector(24.0))), s0(Vector(Vector(25.0), Vector(26.0), Vector(27.0)), Vector(Vector(28.0)), Vector(Vector(29.0))))))
    val v1: Double = 30.0
    val start = nanoTime()
    var result: Double = 30.0
    (1 to 1000).foreach { _ =>
      result = f0(v0, v1)
    }
    val end = nanoTime()
    println(result)
    println(end - start)
  }
  benchmark()
}

File: Benchmarking/src/gen_8_30.swift
  import Dispatch
  struct s0 {
    var p0: [[Double]]
    var p1: [[Double]]
    var p2: [[Double]]
  }
  struct s1 {
    var p0: s0
    var p1: s0
  }
  @inline(never)
  func f0(_ v0: [[s1]], _ v1: Double) -> Double {
    let v7: [s1] = v0[0]
    let v3: s1 = v7[0]
    let v2: s0 = v3.p1
    let v17: [[Double]] = v2.p2
    let v27: [Double] = v17[0]
    let v43: Double = v27[0]
    var v47: Double = v43
    let v91: Double = v47 / v47
    let v64: Double = v27[0]
    let v90: Double = v64 / v91
    v47 = v43
    return v90
  }
  func benchmark() {
    let v0: [[s1]] = [[s1(p0: s0(p0: [[0.0], [1.0], [2.0]], p1: [[3.0]], p2: [[4.0]]), p1: s0(p0: [[5.0], [6.0], [7.0]], p1: [[8.0]], p2: [[9.0]]))], [s1(p0: s0(p0: [[10.0], [11.0], [12.0]], p1: [[13.0]], p2: [[14.0]]), p1: s0(p0: [[15.0], [16.0], [17.0]], p1: [[18.0]], p2: [[19.0]]))], [s1(p0: s0(p0: [[20.0], [21.0], [22.0]], p1: [[23.0]], p2: [[24.0]]), p1: s0(p0: [[25.0], [26.0], [27.0]], p1: [[28.0]], p2: [[29.0]]))]]
    let v1: Double = 30.0
    let start = DispatchTime.now().uptimeNanoseconds
    var result: Double = 30.0
    for _ in 1...1000 {
      result = f0(v0, v1)
    }
    let end = DispatchTime.now().uptimeNanoseconds
    print(result)
    print(end - start)
  }
  benchmark()
